### PR TITLE
feat: Add Azure Cosmos DB (NoSQL) data connector (RC)

### PR DESCRIPTION
**Pull request by lukekim**
## Summary

Adds read-only Azure Cosmos DB (NoSQL / Core SQL API) support as a Spice data connector, at RC quality per [`docs/criteria/connectors/rc.md`](docs/criteria/connectors/rc.md). Built on the official [`azure_data_cosmos`](https://crates.io/crates/azure_data_cosmos) 0.30 SDK.

## RC criteria (from `docs/criteria/connectors/rc.md` row 66)

Per-connector matrix: `Cosmos DB (NoSQL) | ➖ | ➖ | ➖ | ➖ | ➖ | ☑️` — TPC-H / TPC-DS / Federation / Data Correctness / Streaming are intentionally out-of-scope; only inferred schema is required. "All Connectors" gates (resilience, docs, cookbook) are met.

### Connector (alpha → RC)
- `data_components::cosmosdb` module: `CosmosDBClient`, `CosmosDBTableProvider`, `CosmosDBExec` — `TableProvider` + `ExecutionPlan` implementations.
- `runtime::dataconnector::cosmosdb` factory, registered via `register_data_connector!`.
- Arrow schema inferred from a sample of documents; Cosmos system fields (`_rid`, `_self`, `_etag`, `_attachments`, `_ts`) stripped.
- Cargo feature `cosmosdb`, wired through `data_components`, `runtime`, and `spiced`; included in the default `spiced` distribution.

### RC uplift (this branch)
- **Connection resilience:** `CosmosResilienceConfig` (per-endpoint `Semaphore` + `AtomicBool` disabled latch), exponential/fibonacci backoff, `Retry-After` / `x-ms-retry-after-ms` handling, permanent-error (401/403/404) detection.
- **Metrics:** `inflight_operations` gauge via `MetricsProvider`, scoped per dataset connector instance. The shared per-account concurrency budget is enforced separately via an endpoint-keyed `Semaphore` (not this gauge).
- **Unsupported-type handling:** `dataset.unsupported_type_action` honored; all-null inferred columns warn-drop by default.
- **Redacted `Debug`:** `CosmosDBCredential` uses a manual redacted `Debug` impl so account keys never surface in logs.

### Parameters

| Parameter                    | Default       | Description                                                         |
| ---------------------------- | ------------- | ------------------------------------------------------------------- |
| `cosmosdb_connection_string` | —             | Full Cosmos connection string (takes precedence).                   |
| `cosmosdb_account_endpoint`  | —             | Account URL (used with `account_key`).                              |
| `cosmosdb_account_key`       | —             | Primary / secondary key.                                            |
| `cosmosdb_database`          | —             | Overrides the database name parsed from the `from:` path.           |
| `query`                      | `SELECT * FROM c` | Cosmos SQL to scan the container.                                |
| `schema_infer_max_records`   | `100`         | Documents sampled for schema inference.                             |
| `max_concurrent_requests`    | `4`           | Per-account concurrency budget.                                     |
| `http_max_retries`           | `3`           | Retries for transient (429/5xx/network) errors during schema inference. |
| `backoff_method`             | `exponential` | `exponential` or `fibonacci`.                                       |
| `disable_on_permanent_error` | `true`        | Latch the connector disabled on 401/403/404.                        |

## Usage

```yaml
datasets:
  - from: cosmosdb:mydb.mycontainer
    name: my_table
    params:
      cosmosdb_connection_string: ${secrets:cosmosdb_conn}
      # or: cosmosdb_account_endpoint + cosmosdb_account_key
```

Dataset path accepts `database.container`, `database/container`, or just `container` when `cosmosdb_database` is set. See [`examples/cosmosdb-connector/`](examples/cosmosdb-connector/) for a copy-pasteable cookbook recipe.

## What's supported

- Read-only (`SELECT`) cross-partition scans via Cosmos SQL.
- Key-based auth: connection string or account endpoint + account key.
- Arrow schema inference (Utf8 / Int64 / Float64 / Boolean / List / Struct) with JSON → Arrow type table documented at [`docs/dev/cosmosdb.md`](docs/dev/cosmosdb.md#json--arrow-type-mapping).
- Standard Spice acceleration (DuckDB / SQLite / Arrow in-memory) on top of the connector.
- Connection resilience with bounded retry + per-account concurrency budget.

## RC exceptions (documented)

- Filter / projection / limit push-down into Cosmos DB — **post-RC**.
- Writes (`INSERT`/`UPDATE`/`DELETE`) — **post-RC**.
- Change feed streaming (`RefreshMode::Changes`) — **post-RC**.
- Microsoft Entra ID / managed-identity auth — **post-RC**.
- TPC-H / TPC-DS / Federation / Data Correctness / Streaming — exempt per the per-connector RC matrix.
- Retries apply to schema inference and the connect phase; mid-stream pager errors propagate to the refresh layer (which owns its own retry policy).

## Test coverage

- **32** unit tests in `data_components::cosmosdb` (resilience config, backoff math, permanent-error classification, Retry-After parsing, inflight guard, schema inference, JSON→Arrow decoding incl. 1024-column max-width test, unsupported-type actions).
- **12** unit tests in `runtime::dataconnector::cosmosdb` (shared semaphore / disabled flag, path parsing variants).
- **1** offline smoke test + **2** `#[ignore]`'d live tests in `crates/runtime/tests/cosmosdb/`. The live tests read `COSMOSDB_CONNECTION_STRING` (or `COSMOSDB_ACCOUNT_ENDPOINT` + `COSMOSDB_ACCOUNT_KEY`) from the env and run via `cargo test --features cosmosdb -- --ignored cosmosdb_live`.

## Verification

- `cargo check -p data_components --features cosmosdb` — clean.
- `cargo check -p runtime --features cosmosdb` — clean.
- `cargo clippy -p data_components --features cosmosdb -- -D warnings` — clean.
- `cargo clippy -p runtime --features cosmosdb -- -D warnings` — clean.
- `cargo test -p data_components --features cosmosdb --lib cosmosdb` — 32 passed.
- `cargo test -p runtime --features cosmosdb --lib dataconnector::cosmosdb` — 12 passed.

## Notes for reviewers

- Pinned `azure_data_cosmos = 0.30.0` because it aligns with the workspace's existing `azure_core = 0.31.0`. Newer versions pull in `azure_core` 0.32+ and conflict with `runtime`'s dev-deps.
- `azure_data_cosmos` explicitly disables typespec's retry pipeline (`azure_data_cosmos/src/clients/cosmos_client.rs:59`), so the connector owns retry + concurrency limiting itself. Pattern matches the git connector.
- Per-account resilience state (`Semaphore` + disabled `AtomicBool`) is keyed by account endpoint, shared across datasets pointing at the same account. Keys are never evicted during the runtime's lifetime (same trade-off as `GIT_CONCURRENCY_LIMITS`).
- Partition-key strategy is "cross-partition" (`()` arg to `query_items`). Per-partition routing + filter push-down are planned post-RC.

### CODE REVIEW

File: crates/runtime/src/dataconnector/cosmosdb.rs [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        .secret(),
+    ParameterSpec::component("account_key")
+        .description("The Azure Cosmos DB account primary or secondary key.")
+        .secret(),
+    ParameterSpec::component("connection_string")
+        .description("An Azure Cosmos DB connection string (AccountEndpoint=...;AccountKey=...). Takes precedence over account_endpoint/account_key if set.")
+        .secret(),
+    ParameterSpec::component("database")
+        .description("The Cosmos DB database name. Defaults to the first segment of the dataset `from:` path ('database.container').")
+        .required(),
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/runtime/src/dataconnector/cosmosdb.rs]: Fixed in bbbcbab4b — dropped `.required()`. `resolve_database_and_container` already falls back to the first segment of the dataset path, so `from: cosmosdb:mydb.mycontainer` validates without `cosmosdb_database`.

File: crates/runtime/src/dataconnector/cosmosdb.rs [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+            .unwrap_or(DEFAULT_QUERY)
+            .to_string();
+
+        let schema_infer_max_records = self
+            .params
+            .get("schema_infer_max_records")
+            .expose()
+            .ok()
+            .and_then(|v| v.parse::<usize>().ok())
+            .unwrap_or(DEFAULT_SCHEMA_INFER_MAX_RECORDS);
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/runtime/src/dataconnector/cosmosdb.rs]: Fixed in bbbcbab4b — `schema_infer_max_records` now warns and falls back to the default on invalid values (`0`, non-integers), matching the HTTP connector's behavior you referenced.

File: crates/data_components/src/cosmosdb/schema.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        // Caller is expected to map this to a user-facing EmptyContainer
+        // error. Returning an Arrow error here keeps this helper simple.
+        return Ok(Arc::new(Schema::empty()));
+    }
+
+    let schema = infer_json_schema_from_iterator(
+        samples
+            .iter()
+            .map(Result::<_, arrow::error::ArrowError>::Ok),
+    )
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/schema.rs:73]: Fixed in bbbcbab4b — updated the doc comment to match the implementation (returns an empty `Schema` on empty input; callers map that to `EmptyContainer`). The provider already maps empty samples to `EmptyContainer` before inference returns, so the current split is fine for alpha.

File: crates/runtime/src/dataconnector/cosmosdb.rs [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+    // Accept either `database.container` or `database/container`, or just the
+    // container when `database` is explicitly set.
+    let (db_from_path, container) = if let Some((db, container)) = path_str.split_once('.') {
+        (Some(db.to_string()), container.to_string())
+    } else if let Some((db, container)) = path_str.split_once('/') {
+        (Some(db.to_string()), container.to_string())
+    } else {
+        (None, path_str.to_string())
+    };
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/runtime/src/dataconnector/cosmosdb.rs]: Fixed in bbbcbab4b — extracted the pure parsing logic into `parse_database_and_container` and added 9 unit tests covering `db.container`, `db/container`, container-only with explicit `database` param, override behavior, missing database, empty db/container segments, and multiple-separator edge cases. Also added a guard for empty database segments that was silently passing through before.

File: crates/runtime/src/dataconnector/cosmosdb.rs [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+        match (endpoint, key) {
+            (Some(endpoint), Some(key)) => Ok(CosmosDBCredential::Key {
+                endpoint: endpoint.to_string(),
+                key: key.to_string(),
+            }),
+            _ => Err(DataConnectorError::InvalidConfigurationNoSource {
+                dataconnector: CONNECTOR_NAME.to_string(),
+                connector_component: ConnectorComponent::from(dataset),
+                message: "Azure Cosmos DB requires either 'connection_string' or both 'account_endpoint' and 'account_key'.".to_string(),
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/runtime/src/dataconnector/cosmosdb.rs]: Fixed in bbbcbab4b — the error message now uses the user-facing prefixed parameter names (`cosmosdb_connection_string`, `cosmosdb_account_endpoint`, `cosmosdb_account_key`).

File: docs/dev/cosmosdb.md [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+      # Option B — explicit endpoint + key
+      cosmosdb_account_endpoint: https://my-account.documents.azure.com:443/
+      cosmosdb_account_key: ${secrets:cosmosdb_key}
+
+      # Optional: override database (otherwise taken from `from:` path)
+      cosmosdb_database: mydb
+      # Optional: custom Cosmos SQL query (defaults to `SELECT * FROM c`)
+      cosmosdb_query: SELECT * FROM c
+      # Optional: sample size for schema inference (default 100)
+      cosmosdb_schema_infer_max_records: "100"
```
[the remark this answers is not in the document]
Comment (lukekim) [reply docs/dev/cosmosdb.md]: Fixed in bbbcbab4b — docs now use unprefixed `query` and `schema_infer_max_records`, matching their `ParameterSpec::runtime` declarations.

File: crates/data_components/src/cosmosdb/provider.rs [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+    } else {
+        arrow::compute::concat_batches(full_schema, &batches).context(JsonDecodeSnafu)?
+    };
+
+    if let Some(indices) = projection {
+        full_batch.project(indices).context(JsonDecodeSnafu)
+    } else {
+        Ok(full_batch)
+    }
+}
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/provider.rs:546]: Fixed in 7928462e5 — added 4 unit tests for `decode_batch` covering multi-document decoding, column projection, empty-input batches, and missing-field-produces-null behavior. `CosmosDBExec` buffering/streaming is still covered only by the integration layer for now since it requires a live Cosmos client — I've left a note to add a mock-pager test when we revisit for RC.

File: crates/runtime/src/dataconnector/cosmosdb.rs [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+        (None, None) => {
+            return Err(format!(
+                "Could not determine Cosmos DB database from dataset path '{path}'. Expected 'database.container' or set the 'database' parameter."
+            ));
+        }
+    };
+
+    if database.is_empty() {
+        return Err(format!(
+            "Could not determine Cosmos DB database from dataset path '{path}'. Expected 'database.container' or set the 'database' parameter."
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/runtime/src/dataconnector/cosmosdb.rs]: Fixed in 7928462e5 — both error messages in `parse_database_and_container` now say `cosmosdb_database` instead of `database`.

File: crates/data_components/src/cosmosdb/provider.rs [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
+                })?;
+
+            let mut buffer: Vec<Value> = Vec::with_capacity(STREAM_BATCH_SIZE);
+
+            while let Some(item) = pager.next().await {
+                let doc = item.map_err(|e| {
+                    DataFusionError::External(
+                        Box::new(e) as Box<dyn std::error::Error + Send + Sync>
+                    )
+                })?;
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/provider.rs]: Fixed in 7928462e5 — the `execute` async block now wraps `query_items` and per-document pager errors with `Error::QueryFailed { database, container, source }` before converting to `DataFusionError`, matching `fetch_samples`. Runtime errors will now surface the database+container in the message.

File: crates/data_components/src/cosmosdb/provider.rs [hunk 10]
```diff
@@ -0,0 +1,502 @@
...
+) -> Result<RecordBatch, Error> {
+    // `arrow::json::ReaderBuilder` consumes newline-delimited JSON, so join
+    // the document stream with newlines.
+    let mut buf = Vec::with_capacity(docs.len() * 256);
+    for doc in docs {
+        serde_json::to_writer(&mut buf, doc).map_err(|e| Error::JsonDecode {
+            source: arrow::error::ArrowError::JsonError(e.to_string()),
+        })?;
+        buf.push(b'\n');
+    }
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/provider.rs]: Fixed in aab0e70aa — `decode_batch` now uses `ReaderBuilder::build_decoder()` + `decoder.serialize(docs)` + `decoder.flush()` directly, matching the S3 vectors providers. Drops the intermediate `Vec<u8>` buffer and the NDJSON parse pass.

File: docs/dev/cosmosdb.md [hunk 11]
```diff
@@ -0,0 +1,150 @@
+# Azure Cosmos DB (NoSQL / Core SQL) Data Connector
+
+Status: **RC** — read-only scan with RC-level connection resilience.
+
```
[the remark this answers is not in the document]
Comment (lukekim) [reply docs/dev/cosmosdb.md:4]: Fixed — PR title/body now explicitly state RC status (see the updated description). Cosmos DB meets the per-connector RC matrix row 66 (TPC-H/DS/Federation/Data Correctness/Streaming all `➖`; inferred schema `☑️`) plus the mandatory "All Connectors" gates (connection resilience, unsupported-type warn-and-skip, docs with type map + RC exceptions + cookbook).

File: docs/criteria/connectors/rc.md [hunk 12]
```diff
@@ -11,6 +11,7 @@ All criteria must be met for the connector to be considered [RC](../definitions.
 | ADBC                             | ➖          |              |
 | Azure BlobFS                     | ➖          |              |
 | Clickhouse                       | ➖          |              |
+| Cosmos DB (NoSQL)                | ✅          | @lukekim     |
 | Databricks (mode: delta_lake)    | ✅          | @Sevenannn   |
 | Databricks (mode: spark_connect) | ➖          |              |
```
[the remark this answers is not in the document]
Comment (lukekim) [reply docs/criteria/connectors/rc.md:16]: Fixed — PR title/body now explicitly state RC status (see the updated description). Cosmos DB meets the per-connector RC matrix row 66 (TPC-H/DS/Federation/Data Correctness/Streaming all `➖`; inferred schema `☑️`) plus the mandatory "All Connectors" gates (connection resilience, unsupported-type warn-and-skip, docs with type map + RC exceptions + cookbook).

File: docs/criteria/connectors/beta.md [hunk 13]
```diff
@@ -11,6 +11,7 @@ All criteria must be met for the connector to be considered Beta, with exception
 | ADBC                             | ➖           |                 |
 | Azure BlobFS                     | ➖           |                 |
 | Clickhouse                       | ➖           |                 |
+| Cosmos DB (NoSQL)                | ✅           | @lukekim        |
 | Databricks (mode: delta_lake)    | ✅           | @Sevenannn      |
 | Databricks (mode: spark_connect) | ✅           | @Sevenannn      |
```
[the remark this answers is not in the document]
Comment (lukekim) [reply docs/criteria/connectors/beta.md:16]: Fixed — PR title/body now explicitly state RC status (see the updated description). Cosmos DB meets the per-connector RC matrix row 66 (TPC-H/DS/Federation/Data Correctness/Streaming all `➖`; inferred schema `☑️`) plus the mandatory "All Connectors" gates (connection resilience, unsupported-type warn-and-skip, docs with type map + RC exceptions + cookbook).

File: crates/data_components/src/cosmosdb/provider.rs [hunk 14]
```diff
@@ -0,0 +1,724 @@
...
+                        source: Box::new(err),
+                    })
+                };
+
+            let mut pager = container_client
+                .query_items::<Value>(config.query.as_str(), (), None)
+                .map_err(|e| handle_stream_error(&config.resilience, client.endpoint(), e))?;
+
+            let mut buffer: Vec<Value> = Vec::with_capacity(STREAM_BATCH_SIZE);
+
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/provider.rs:490]: Documented in 1cb0e22f7 — `docs/dev/cosmosdb.md` now explicitly calls out that `http_max_retries` / `backoff_method` apply to the schema-inference pass at dataset registration. Mid-stream pager errors propagate directly to the caller because a `FeedPager` cannot be rewound once rows have been emitted without risking duplicate output; retry is owned by Spice's dataset refresh layer at the query boundary. The permanent-error latch (`disable_on_permanent_error`) still fires on both paths so a 401/403/404 anywhere disables the connector account-wide.

File: crates/data_components/src/cosmosdb/mod.rs [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+*/
+
+//! Azure Cosmos DB (`NoSQL` / Core SQL API) data connector components.
+//!
+//! Exposes a [`provider::CosmosDBTableProvider`] built on top of the
+//! `azure_data_cosmos` crate. Documents are fetched via a Cosmos SQL query
+//! (`SELECT * FROM c` by default) and projected into Arrow `RecordBatch`es.
+//!
+//! The first pass targets *alpha* quality: read-only, cross-partition scan,
+//! schema inference from a sample of documents, no filter push-down yet.
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/mod.rs]: Fixed — crate-level docs in `crates/data_components/src/cosmosdb/mod.rs` now state *RC* quality and cross-reference the RC matrix row and dev doc for the full type map + limitations.

File: crates/runtime/src/dataconnector/cosmosdb.rs [hunk 16]
```diff
@@ -0,0 +1,577 @@
...
+    }
+}
+
+#[derive(Debug)]
+pub struct CosmosDB {
+    params: Parameters,
+    /// Drives the `inflight_operations` metric gauge. Cloned into every
+    /// [`CosmosResilienceConfig`] this connector produces so the gauge
+    /// aggregates across datasets targeting the same Cosmos account.
+    inflight_operations: Arc<AtomicU64>,
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/runtime/src/dataconnector/cosmosdb.rs]: Fixed — updated the doc comment in `CosmosDB` to accurately describe semantics: the counter is per-dataset (one `CosmosDB` per dataset), and the shared per-endpoint concurrency budget is enforced via the endpoint-keyed `COSMOS_CONCURRENCY_LIMITS` map, not this metric. Keying the gauge by endpoint is a behavior change we can revisit separately; for RC the accurate-docs path is the right scope.

File: crates/runtime/src/dataconnector/cosmosdb.rs [hunk 17]
```diff
@@ -0,0 +1,577 @@
...
+        .default("100"),
+
+    ParameterSpec::runtime("max_concurrent_requests")
+        .description("Maximum number of concurrent Azure Cosmos DB requests per account endpoint, shared across all datasets pointing at the same account.")
+        .default("4"),
+    ParameterSpec::runtime("http_max_retries")
+        .description("Maximum number of retries for transient errors (429, 5xx, network) when reading from Azure Cosmos DB. Retries use the configured backoff strategy and honor Retry-After headers.")
+        .default("3"),
+    ParameterSpec::runtime("backoff_method")
+        .description("Backoff strategy between retries on transient errors. 'exponential' doubles the delay each attempt; 'fibonacci' follows the Fibonacci sequence.")
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/runtime/src/dataconnector/cosmosdb.rs]: Fixed — `http_max_retries` and `backoff_method` descriptions now explicitly state they apply to the schema-inference sampling pass at dataset registration, and call out that mid-stream pager errors during scan execution are not retried. This matches the implementation (only `fetch_samples` uses `run_with_resilience`) and the guidance already documented in `docs/dev/cosmosdb.md`.

File: crates/data_components/src/cosmosdb/provider.rs [hunk 18]
```diff
@@ -0,0 +1,711 @@
...
+//! The provider executes a single Cosmos SQL query (defaulting to
+//! `SELECT * FROM c`) against the configured container, infers an Arrow
+//! schema from a sample of documents on first access, and streams the full
+//! result set into record batches via `arrow::json::ReaderBuilder`.
+//!
+//! This is an alpha-quality implementation:
+//! * Read-only (no INSERT / UPDATE / DELETE).
+//! * Cross-partition scan only — no filter or projection push-down.
+//! * Schema inferred from a sample; pin the schema via the dataset
+//!   `columns:` spicepod property when stability is required.
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/provider.rs]: Fixed — module-level docs in `provider.rs` now state RC quality and include the retry-scope limitation ("retries/backoff apply to the schema-inference pass only; mid-stream pager errors during scan execution propagate directly") alongside the other current limitations.

**Review by sgrebnov — DISMISSED** [review]

File: docs/dev/cosmosdb.md [hunk 19]
```diff
@@ -0,0 +1,159 @@
...
+- Arrow schema inferred from a sample of documents (system fields
+  `_rid`, `_self`, `_etag`, `_attachments`, `_ts` are stripped). Pin a schema
+  via the dataset `columns:` property when stability is required.
+- Standard Spice acceleration (DuckDB / SQLite / Arrow in-memory) on top of
+  the connector.
+- Connection resilience: per-account concurrency semaphore, bounded retries
+  with configurable backoff, `Retry-After` / `x-ms-retry-after-ms` handling,
+  permanent-error (401/403/404) detection that latches the connector disabled.
+- `inflight_operations` metric gauge, exported via the runtime metrics
+  endpoint for dashboards.
```
[the remark this answers is not in the document]
Comment (lukekim) [reply docs/dev/cosmosdb.md]: Fixed — `docs/dev/cosmosdb.md` now spells out that the `inflight_operations` gauge is scoped per dataset connector instance, not an account-wide aggregate. This matches the per-dataset semantics we also called out in the `CosmosDB` doc comment in the previous commit.

File: crates/data_components/src/cosmosdb/client.rs [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+        })
+    }
+
+    #[must_use]
+    pub fn endpoint(&self) -> &str {
+        &self.endpoint
+    }
+
+    #[must_use]
+    pub fn container_client(&self, database: &str, container: &str) -> ContainerClient {
```
Comment (krinart) [inline crates/data_components/src/cosmosdb/client.rs]: Nut sure I understand why we need this method since every `CosmosDBTableProvider` is tied to a specific combination of `(database, container)`.

Which means we can:
* Remove `CosmosDBClient`
* Initialize `ContainerClient` and keep reusing it.

Comment (lukekim) [reply crates/data_components/src/cosmosdb/client.rs]: Good catch — refactored in 7e574622a. Deleted the `CosmosDBClient` wrapper entirely and replaced it with a free `build_container_client(credential, database, container) -> (ContainerClient, Arc<str> /* endpoint */)` helper. The `ContainerClient` is now built once at connector setup and stored directly on `CosmosDBTableProvider` and `CosmosDBExec`, so `fetch_samples` and the `execute` path reuse it with no per-call `.container_client(db, c)` indirection. Endpoint is kept alongside as `Arc<str>` since it's still needed for resilience keying and error context.

**Review by krinart — DISMISSED** [review]

File: examples/cosmosdb-connector/spicepod.yaml [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+  # Pinned schema — avoids inference drift when optional fields are null in
+  # the first N sampled documents but not in general.
+  - from: cosmosdb:store.products
+    name: products_pinned
+    description: Products with an explicit Arrow schema.
+    params:
+      cosmosdb_connection_string: ${ env:COSMOSDB_CONNECTION_STRING }
+      # Resilience tuning: tighten the per-account concurrency budget when
+      # the dataset is expected to run under heavy load.
+      max_concurrent_requests: "8"
```
[the remark this answers is not in the document]
Comment (lukekim) [reply examples/cosmosdb-connector/spicepod.yaml:33]: Fixed — updated the example comment to say "raise" instead of "tighten" and explicitly note the default (4), so the direction now matches the demonstrated value (8). Kept the value at 8 because the rest of the example (`http_max_retries: 5`, `backoff_method: fibonacci`) is demonstrating a more-aggressive resilience profile for heavy-load datasets rather than a conservative one.

**Review by krinart — DISMISSED** [review]

File: docs/dev/cosmosdb.md [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+  via the dataset `columns:` property when stability is required.
+- Standard Spice acceleration (DuckDB / SQLite / Arrow in-memory) on top of
+  the connector.
+- Connection resilience: per-account concurrency semaphore, bounded retries
+  with configurable backoff, `Retry-After` / `x-ms-retry-after-ms` handling,
+  permanent-error (401/403/404) detection that latches the connector disabled.
+- `inflight_operations` metric gauge, exported via the runtime metrics
+  endpoint for dashboards. This gauge is scoped per dataset connector
+  instance, not as an account-wide aggregate across all datasets using the
+  same Cosmos account endpoint.
```
[the remark this answers is not in the document]
Comment (lukekim) [reply docs/dev/cosmosdb.md:58]: Fixed — updated the PR description so the `inflight_operations` bullet matches the code and docs: "scoped per dataset connector instance. The shared per-account concurrency budget is enforced separately via an endpoint-keyed `Semaphore` (not this gauge)." The docs (`docs/dev/cosmosdb.md` line 55–58) and the `CosmosDB` struct doc comment already call this out; this just aligns the PR description.

**Review by krinart — DISMISSED** [review]

File: crates/data_components/src/cosmosdb/client.rs [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+        }
+        CosmosDBCredential::Key { endpoint, key } => {
+            let client = CosmosClient::with_key(&endpoint, Secret::from(key), None)
+                .map_err(boxed_err)
+                .context(BuildClientSnafu {
+                    endpoint: endpoint.clone(),
+                })?;
+
+            (client, endpoint)
+        }
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/client.rs:96]: Fixed — added `normalize_endpoint` (trim trailing `/`, trim whitespace, lowercase) and applied it to the endpoint before returning it from `build_container_client`. Benign URL-formatting variants (trailing slash, casing) now collapse to the same shared-semaphore key, so datasets targeting the same account share the concurrency budget regardless of configured spelling. Added a unit test that pins the normalization for a few common variants.

File: crates/data_components/src/cosmosdb/resilience.rs [hunk 24]
```diff
@@ -0,0 +1,619 @@
...
+        // Large attempt saturates at the cap.
+        assert_eq!(
+            backoff_delay(BackoffMethod::Exponential, 100),
+            RETRY_MAX_BACKOFF
+        );
+    }
+
+    #[test]
+    fn backoff_delay_fibonacci_grows_as_expected() {
+        // Factors follow F(attempt+1) with F(1)=F(2)=1 → 1, 2, 3, 5, 8, ...
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/resilience.rs]: Fixed — updated the comment to `F(attempt+2)` with F(1)=F(2)=1. Traced through: `backoff_delay` starts with `a=1, b=1` and returns `b` after `attempt` iterations of `(next=a+b; a=b; b=next)`, so attempts 0..4 map to 1, 2, 3, 5, 8 — matching F(attempt+2). The sequence and assertions were always correct; only the comment was off by one.

**Review by krinart — DISMISSED** [review]

File: examples/cosmosdb-connector/spicepod.yaml [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+      - name: id
+        type: string
+      - name: name
+        type: string
+      - name: price
+        type: double
+      - name: tags
+        type: list<string>
+      - name: available
+        type: boolean
```
[the remark this answers is not in the document]
Comment (lukekim) [reply examples/cosmosdb-connector/spicepod.yaml]: Good catch — you're right, Spicepod's `Column` is semantic-only (no `type` field), so the `type:` keys were silently ignored and the `products_pinned` dataset wasn't actually pinning anything. Rewrote that example as a resilience-tuning demo (`max_concurrent_requests`, `http_max_retries`, `backoff_method`) and widened `schema_infer_max_records` to stabilize inference when optional fields are sparse — that's the real lever we have today.

File: docs/dev/cosmosdb.md [hunk 26]
```diff
@@ -0,0 +1,161 @@
...
+| object `{ ... }`            | `Struct`        | Nested objects are preserved as structs.                                                                                                                    |
+| array `[ ... ]`             | `List`          | The element type is inferred from the first non-null item; heterogeneous arrays may require `columns:` pinning to disambiguate.                             |
+| all-null in sample          | `Null`          | Warn-dropped by default (`unsupported_type_action=warn`). Pin via `columns:` to force a type, or set `unsupported_type_action=string` to coerce to `Utf8`.  |
+| System fields (`_rid`, ...) | stripped        | Never appear in the dataset schema.                                                                                                                         |
+
+Cosmos does not emit `Date`, `Time`, `Timestamp`, `Decimal`, or `Binary`
+natively — they round-trip as strings. Pin those with explicit `columns:`
+declarations and use `CAST` at query time.
```
[the remark this answers is not in the document]
Comment (lukekim) [reply docs/dev/cosmosdb.md]: Fixed — `docs/dev/cosmosdb.md` no longer claims schema pinning via `columns:`. The "What's supported" bullet now says "Schema pinning is not currently supported — widen `schema_infer_max_records` instead to stabilize inference when optional fields are sparse", and the type-map rows for heterogeneous arrays / all-null samples / Date-Time-Decimal-Binary were rewritten to recommend the actually-supported levers (widen the sample, use `unsupported_type_action=string`, CAST at query time).

File: crates/data_components/src/cosmosdb/provider.rs [hunk 27]
```diff
@@ -0,0 +1,736 @@
...
+        self.schema_infer_max_records = n;
+        self
+    }
+
+    #[must_use]
+    pub fn with_schema_override(mut self, schema: SchemaRef) -> Self {
+        self.schema_override = Some(schema);
+        self
+    }
+
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/data_components/src/cosmosdb/provider.rs]: Removed — deleted the unused `schema_override` field and `with_schema_override` builder from `CosmosDBTableProviderConfig`, and simplified `CosmosDBTableProvider::try_new` to always infer schema (no more dead "if override else infer" branch). Docs/example were updated in the same commit to stop claiming a pinning path that didn't exist. All 33 cosmosdb unit tests still pass.

File: crates/runtime/src/dataconnector/cosmosdb.rs [hunk 28]
```diff
@@ -0,0 +1,592 @@
...
+            .or_insert_with(|| Arc::new(AtomicBool::new(false))),
+    )
+}
+
+const COSMOSDB_METRICS: &[MetricSpec] =
+    &[
+        MetricSpec::new("inflight_operations", MetricType::ObservableGaugeU64)
+            .description("Current number of Azure Cosmos DB requests holding a concurrency permit")
+            .auto_register(),
+    ];
```
[the remark this answers is not in the document]
Comment (lukekim) [reply crates/runtime/src/dataconnector/cosmosdb.rs:116]: Fixed by clarifying the description — `inflight_operations` now reads "Azure Cosmos DB operations currently holding a concurrency permit — incremented once per operation and held across retry backoff sleeps (not a pure in-flight-HTTP counter)". This matches the actual semantics (counter + semaphore permit are both held for the full duration of an operation including backoff waits). Adjusting the implementation to track only active HTTP requests would mean releasing/reacquiring across retries, which would also defeat the concurrency budget enforcement — so the description route is the right fix here.

**Review by krinart — DISMISSED** [review]

**Review by krinart — APPROVED** [review]
